docs/tutorial/flow-nft-marketplace.md

```diff
@@ -0,0 +1,1804 @@
+
+Since there will be quite a lot going on, we will go slowly on this one. We will create a `mintToken` function that takes a `pet` object and does the following:
+
+1. Upload the metadata and image asset to NFT.storage, and retrieve the returned metadata that includes the [CID][ipfs-cid] of the data.
+2. Send a minting transaction with the metadata to Flow (in this case, the name, age, breed, and the CID of the data stored on IPFS).
+3. If successful, return the Flow transaction ID.
+
+
+Here we sketch up some placeholder functions to reflect the steps:
+
+```js
+
+// MintToken.tx.js
+
+async function mintToken(pet) {
+  let metadata = await uploadToStorage(pet);
+  let txId = await mintPet(metadata);
+  return txId;
+}
+
+/* We will fill in these functions next */
+
+async function uploadToStorage(pet) {
+    return {};
+}
+
+async function mintPet(metadata) {
+    return '';
+}
+
+```
+
+> **💡 Where to catch an error?**
+> Asynchronous functions which return a `Promise` are *fallable*. Conventionally, you should prepare for the worst
+> by catching the error. But the question is where to do this?
+> The general rule here is if a function does not know what to do with the error, don't catch it. Delegate this responsibility
+> to the caller who, hopefully, will know better. In the previous `mintToken` function, we didn't
+> handle the error because we wanted the calling function to handle it.
+> This way, you are not making things worse by complicating the code with unnecessary `try-catch` blocks that only obscure
+> the errors. Let the one who knows better deals with it.
+
+Next, we will fill in the body of `uploadToStorage` function. This is where you will need your API key from NFT.Storage:
+
+```js
+
+// Import these modules from nft.storage on top of the file.
+import { NFTStorage, File } from 'nft.storage';
+
+const API_KEY = "DROP_YOUR_API_KEY_HERE";
+
+// Initialize the NFTStorage client.
+const storage = new NFTStorage({ token: API_KEY });
+
+async function uploadToStorage(pet) {
+  // Call `store(...)` on the NFTStorage client with an object
+  // containing all of pet's attributes, and required image and
+  // description attributes.
+  let metadata = await storage.store({
+    ...pet,
+    image: pet.image && new File([pet.image], `${pet.name}.jpg`, { type: 'image/jpg' }),
+    description: `${pet.name}'s metadata`,
+  });
+
+  // If all goes well, return the metadata.
+  return metadata;
+}
+
+```
+
+The step was simple. `NFTStorage.store(...)` takes an object with arbitrary attributes and two required `image` and `description` attributes.
+
+Confusingly, `image` attribute does not necessarily take only an image file. It takes a `File` object (which means any type of assets). We used the `File` contructor imported from the library to create the object.
+
+The `description` attribute can obvious be any arbitrary text you want.
```

Comment:
```suggestion
The `description` attribute can be any arbitrary text up to $MAXLENGTH.
```
@yusefnapora @dchoi27  is there a max length, size, or other restrictions on `description` attribute?